src/wallet/wallet2.cpp

```diff
@@ -7917,6 +7917,198 @@ bool wallet2::check_tx_proof(const crypto::hash &txid, const cryptonote::account
+    "Signature decoding error");
+
+  std::istringstream iss(sig_decoded);
+  boost::archive::portable_binary_iarchive ar(iss);
+  std::vector<reserve_proof_entry> proofs;
+  ar >> proofs;
+
+  // fetch txes from daemon
+  COMMAND_RPC_GET_TRANSACTIONS::request gettx_req;
+  COMMAND_RPC_GET_TRANSACTIONS::response gettx_res;
+  for (size_t i = 0; i < proofs.size(); ++i)
+    gettx_req.txs_hashes.push_back(epee::string_tools::pod_to_hex(proofs[i].txid));
+  m_daemon_rpc_mutex.lock();
+  bool ok = net_utils::invoke_http_json("/gettransactions", gettx_req, gettx_res, m_http_client);
+  m_daemon_rpc_mutex.unlock();
+  THROW_WALLET_EXCEPTION_IF(!ok || (gettx_res.txs.size() != proofs.size() && gettx_res.txs_as_hex.size() != proofs.size()),
+    error::wallet_internal_error, "Failed to get transaction from daemon");
+
+  // check spent status
+  COMMAND_RPC_IS_KEY_IMAGE_SPENT::request kispent_req;
+  COMMAND_RPC_IS_KEY_IMAGE_SPENT::response kispent_res;
+  for (size_t i = 0; i < proofs.size(); ++i)
+    kispent_req.key_images.push_back(epee::string_tools::pod_to_hex(proofs[i].ki));
+  m_daemon_rpc_mutex.lock();
+  ok = epee::net_utils::invoke_http_json("/is_key_image_spent", kispent_req, kispent_res, m_http_client, rpc_timeout);
+  m_daemon_rpc_mutex.unlock();
+  THROW_WALLET_EXCEPTION_IF(!ok || kispent_res.spent_status.size() != proofs.size(),
+    error::wallet_internal_error, "Failed to get key image spent status from daemon");
+
+  total = spent = 0;
+  auto tx_as_hex = gettx_res.txs_as_hex.begin();
+  for (size_t i = 0; i < proofs.size(); ++i)
+  {
+    cryptonote::blobdata tx_data;
+    if (gettx_res.txs.size() == proofs.size())
+      ok = string_tools::parse_hexstr_to_binbuff(gettx_res.txs[i].as_hex, tx_data);
+    else
+      ok = string_tools::parse_hexstr_to_binbuff(*tx_as_hex++, tx_data);
+    THROW_WALLET_EXCEPTION_IF(!ok, error::wallet_internal_error, "Failed to parse transaction from daemon");
+
+    crypto::hash tx_hash, tx_prefix_hash;
+    cryptonote::transaction tx;
+    THROW_WALLET_EXCEPTION_IF(!cryptonote::parse_and_validate_tx_from_blob(tx_data, tx, tx_hash, tx_prefix_hash), error::wallet_internal_error,
+      "Failed to validate transaction from daemon");
+    THROW_WALLET_EXCEPTION_IF(tx_hash != proofs[i].txid, error::wallet_internal_error, "Failed to get the right transaction from daemon");
+
+    bool found = false;
+    for (size_t n = 0; n < tx.vout.size(); ++n)
+    {
+      const cryptonote::txout_to_key* const out_key = boost::get<cryptonote::txout_to_key>(std::addressof(tx.vout[n].target));
+      if (!out_key || proofs[i].key != out_key->key)
+        continue;
+      found = true;
+
+      std::string prefix_data((const char*)&proofs[i].txid, sizeof(crypto::hash));
+      prefix_data += std::string((const char*)&proofs[i].key, sizeof(crypto::public_key));
+      prefix_data += message;
+      crypto::hash prefix_hash;
```

Comment:
I think a static string should be appended after the message. Something to make it impossible for the prover to sign a spendable transaction.

One thing I do not like about this approach is that there is no actual proof that a specific address owns the funds. A man in the middle type of attack could be done if the verifier allows a poor message to be used (i.e. I could post someone elses verification in the worse case). I guess there is no way around this, except for some good information in the help documentation?